File: src/main/java/com/example/the_machine/service/RunService.java
package com.example.the_machine.service;

import static com.example.the_machine.db.AssistantEntity.DEFAULT_ASSISTANT_ID;

import com.example.the_machine.db.ArtifactEntity;
import com.example.the_machine.db.ArtifactEntity.Status;
import com.example.the_machine.db.RunEntity;
import com.example.the_machine.domain.event.ArtifactCreatedEvent;
import com.example.the_machine.domain.event.ArtifactScheduledEvent;
import com.example.the_machine.domain.event.RunCompletedEvent;
import com.example.the_machine.domain.event.RunScheduledEvent;
import com.example.the_machine.domain.event.RunStartedEvent;
import com.example.the_machine.dto.CreateRunRequestDto;
import com.example.the_machine.dto.RunDto;
import com.example.the_machine.mapper.RunMapper;
import com.example.the_machine.repo.ArtifactRepository;
import com.example.the_machine.repo.RunRepository;
import com.example.the_machine.repo.ThreadRepository;
import com.example.the_machine.service.events.EventService;
import com.example.the_machine.service.events.SseHub;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

@Service
@RequiredArgsConstructor
@Slf4j
public class RunService {

  private final RunRepository runRepository;
  private final ThreadRepository threadRepository;
  private final RunMapper runMapper;
  private final EventService eventService;
  private final ExecutorService runExecutorService;
  private final ArtifactRepository artifactRepository;
  private final SseHub sseHub;
  private final ObjectMapper objectMapper;

  @Transactional
  public RunDto scheduleRun(CreateRunRequestDto request) {
    final var thread = threadRepository.findById(request.threadId())
        .orElseThrow(() -> new IllegalArgumentException("Thread not found: " + request.threadId()));
    final var threadId = thread.getId();
    log.info("Creating run for thread: {}", threadId);

    final var run = runRepository.save(RunEntity.builder()
        .threadId(threadId)
        .assistantId(DEFAULT_ASSISTANT_ID)
        .task(request.task())
        .status(RunEntity.Status.QUEUED)
        .build()
    );

    eventService.publish(new RunScheduledEvent(threadId, run.getId()));
    log.info("Run created and queued: {}", run.getId());
    return runMapper.toDto(run);
  }

  @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
  public void handleRunCreated(RunScheduledEvent event) {
    log.debug("Handling run created event for run: {}", event.getPayload().runId());
    runExecutorService.submit(() -> executeRun(event.getPayload().runId()));
  }

  public void executeRun(UUID runId) {
    log.info("Starting execution of run: {}", runId);

    final var run = runRepository.findById(runId).orElseThrow();
    run.start();
    eventService.publish(new RunStartedEvent(run.getThreadId(), runId));

    final var threadId = run.getThreadId();

    CompletableFuture.runAsync(() -> {
      try {
        // ==================== TEXT ARTIFACT ====================
        log.info("Creating text artifact...");

        var textArtifact = ArtifactEntity.builder()
            .threadId(threadId)
            .runId(runId)
            .kind(ArtifactEntity.Kind.STRUCTURED_JSON)
            .status(Status.SCHEDULED)
            .title("Brand Analysis Report")
            .mime("application/json")
            .build();

        textArtifact = artifactRepository.save(textArtifact);
        final var textArtifactId = textArtifact.getId();
        log.info("Created text artifact: {}", textArtifactId);

        // Minimal event - just the ID
        sseHub.broadcast(new ArtifactScheduledEvent(threadId, textArtifactId));
        log.info("Broadcasted artifact.scheduled for {}", textArtifactId);

        Thread.sleep(1500);

        // Create content
        var analysisData = objectMapper.createObjectNode()
            .put("brandName", "Nike")
            .put("sentiment", "Very Positive")
            .put("score", 92)
            .put("summary",
                "Nike demonstrates strong brand equity with innovative products and emotional marketing.")
            .set("keywords", objectMapper.createArrayNode()
                .add("innovation")
                .add("athletics")
                .add("inspiration")
                .add("performance")
            );

        textArtifact.setContent(analysisData);
        textArtifact.setStatus(Status.CREATED);
        artifactRepository.save(textArtifact);
        log.info("Updated text artifact to CREATED");

        // Minimal event - frontend will fetch details
        sseHub.broadcast(new ArtifactCreatedEvent(threadId, textArtifactId));
        log.info("Broadcasted artifact.created for {}", textArtifactId);

        Thread.sleep(1000);

        // ==================== IMAGE ARTIFACT ====================
        log.info("Creating image artifact...");

        var imageArtifact = ArtifactEntity.builder()
            .threadId(threadId)
            .runId(runId)
            .kind(ArtifactEntity.Kind.IMAGE)
            .status(Status.SCHEDULED)
            .title("Nike Campaign Hero Image")
            .mime("image/png")
            .build();

        imageArtifact = artifactRepository.save(imageArtifact);
        final var imageArtifactId = imageArtifact.getId();
        log.info("Created image artifact: {}", imageArtifactId);

        sseHub.broadcast(new ArtifactScheduledEvent(threadId, imageArtifactId));
        log.info("Broadcasted artifact.scheduled for {}", imageArtifactId);

        Thread.sleep(2000);

        // Update with storage key
        String storageKey = "066cad46-e944-4247-8ab4-4fa3712cb16e.png";
        imageArtifact.setStorageKey(storageKey);
        imageArtifact.setStatus(Status.CREATED);

        var metadata = objectMapper.createObjectNode()
            .put("size", 2_655_864)
            .put("width", 1024)
            .put("height", 1536)
            .put("format", "png");
        imageArtifact.setMetadata(metadata);

        artifactRepository.save(imageArtifact);
        log.info("Updated image artifact to CREATED with storage key: {}", storageKey);

        sseHub.broadcast(new ArtifactCreatedEvent(threadId, imageArtifactId));
        log.info("Broadcasted artifact.created for {}", imageArtifactId);

        log.info("Artifact simulation completed successfully!");

      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        log.error("Artifact simulation interrupted", e);
      } catch (Exception e) {
        log.error("Error during artifact simulation", e);
      }
    });

    run.complete("Done");
    runRepository.save(run);
    eventService.publish(new RunCompletedEvent(run.getThreadId(), runId));

    log.info("Run execution completed: {}", runId);
  }
}

File: src/main/kotlin/com/example/the_machine/service/RunService.kt
package com.example.the_machine.service

import com.example.the_machine.common.KotlinSerializationConfig
import com.example.the_machine.domain.AssistantEntity
import com.example.the_machine.domain.EventType
import com.example.the_machine.domain.RunEntity
import com.example.the_machine.dto.CreateRunRequestDto
import com.example.the_machine.dto.EventEnvelopeDto
import com.example.the_machine.dto.RunDto
import com.example.the_machine.repo.RunRepository
import com.example.the_machine.repo.ThreadRepository
import com.example.the_machine.service.events.EventPublisher
import com.example.the_machine.service.events.RunCreatedEvent
import com.example.the_machine.service.mapping.RunMapper
import com.example.the_machine.service.mapping.ThreadMapper
import kotlinx.serialization.json.encodeToJsonElement
import mu.KotlinLogging
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener
import java.time.Instant
import java.util.*
import java.util.concurrent.ExecutorService

private val log = KotlinLogging.logger {}

/**
 * Service for managing run creation and execution.
 */
@Service
class RunService(
  private val runRepository: RunRepository,
  private val threadRepository: ThreadRepository,
  private val runMapper: RunMapper,
  private val threadMapper: ThreadMapper,
  private val eventPublisher: EventPublisher,
  private val runExecutorService: ExecutorService,
  private val assistantEngine: AssistantEngine,
  private val applicationEventPublisher: ApplicationEventPublisher,
  private val artifactService: ArtifactService
) {

  /**
   * Creates a new run based on the request.
   *
   * @param request the create run request
   * @return the created run DTO
   */
  @Transactional
  fun createRun(request: CreateRunRequestDto): RunDto {
    log.info { "Creating run for thread: ${request.threadId}" }

    // Load thread
    val thread = threadRepository.findById(request.threadId)
      .orElseThrow { IllegalArgumentException("Thread not found: ${request.threadId}") }

    // Note: MessageService would handle user message creation if available
    // For now, we just log that input was provided
    if (request.input != null) {
      log.debug { "User input provided for run" }
    }

    // Create run entity directly
    val run = RunEntity(
      threadId = request.threadId,
      assistantId = request.assistantId ?: AssistantEntity.DEFAULT_ASSISTANT_ID,
      status = RunEntity.Status.QUEUED,
      kind = RunEntity.Kind.FULL,
    )

    val savedRun = runRepository.save<RunEntity>(run)
    val runDto = runMapper.toDto(savedRun)

    // Log run creation event
    val envelope = EventEnvelopeDto(
      EventType.RUN_CREATED,
      thread.id,
      savedRun.id,
      null,
      KotlinSerializationConfig.staticJson.encodeToJsonElement(runDto)
    )
    eventPublisher.publishAndStore(thread.id, envelope, null)

    // Publish event for async execution after transaction commit
    applicationEventPublisher.publishEvent(RunCreatedEvent(savedRun.id, thread.id))

    log.info { "Run created and queued: ${savedRun.id}" }
    return runDto
  }

  /**
   * Handles run created events after transaction commit to ensure async execution starts only when
   * the run is persisted.
   *
   * @param event the run created event
   */
  @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
  fun handleRunCreated(event: RunCreatedEvent) {
    log.debug { "Handling run created event for run: ${event.runId}" }
    runExecutorService.submit { executeRun(event.runId, event.threadId) }
  }

  /**
   * Executes a run in the background.
   *
   * @param runId the ID of the run to execute
   * @param threadId the ID of the thread associated with the run
   */
  private fun executeRun(runId: UUID, threadId: UUID) {
    log.info { "Starting execution of run: $runId" }

    try {
      // Always fetch fresh managed entity by ID
      val currentRun = runRepository.findById(runId)
        .orElseThrow { IllegalStateException("Run not found: $runId") }

      // Transition to RUNNING
      val updatedRun = currentRun.copy(
        status = RunEntity.Status.RUNNING,
        startedAt = Instant.now()
      )
      val run = runRepository.save(updatedRun)

      val envelope = EventEnvelopeDto(
        EventType.RUN_UPDATED,
        threadId,
        runId,
        null,
        KotlinSerializationConfig.staticJson.encodeToJsonElement(runMapper.toDto(run))
      )
      eventPublisher.publishAndStore(threadId, envelope, null)

      // Create run context with IDs
      val context = RunContext(
        threadId,
        runId,
        runRepository,
        threadRepository,
        runMapper,
        threadMapper,
        eventPublisher,
        artifactService,
        RunPolicy.defaultPolicy()
      )

      // Simple planner logic
      val executionType = determineExecutionType(run)

      when (executionType) {
        ExecutionType.PLAN_CLARIFY -> assistantEngine.executePlanAskClarify(context)
        ExecutionType.FULL_PIPELINE -> assistantEngine.executeFullPipeline(context)
        ExecutionType.PARTIAL_REVISION -> assistantEngine.executePartialRevision(
          context,
          run.reason ?: ""
        )
      }

      log.info { "Run execution completed: $runId" }

    } catch (e: Exception) {
      log.error(e) { "Run execution failed: $runId" }

      // Always fetch fresh managed entity by ID for error handling
      val currentRun = runRepository.findById(runId)
        .orElseThrow { IllegalStateException("Run not found: $runId") }

      val failedRun = currentRun.copy(
        status = RunEntity.Status.FAILED,
        error = e.message,
        completedAt = Instant.now()
      )
      val run = runRepository.save(failedRun)

      val errorEnvelope = EventEnvelopeDto(
        EventType.RUN_UPDATED,
        threadId,
        runId,
        null,
        KotlinSerializationConfig.staticJson.encodeToJsonElement(runMapper.toDto(run))
      )
      eventPublisher.publishAndStore(threadId, errorEnvelope, null)
    }
  }

  /**
   * Determines the execution type based on run and thread state. Simplified logic for MVP without
   * complex message checking.
   *
   * @param run the current run
   * @return the execution type
   */
  private fun determineExecutionType(run: RunEntity): ExecutionType {
    // Check if reason indicates partial revision
    val reason = run.reason
    if (reason != null && reason.startsWith("user_revision:")) {
      return ExecutionType.PARTIAL_REVISION
    }

    // Simple logic for MVP: assume FULL kind means full pipeline
    // In a real implementation, this would check message history and context
    if (run.kind == RunEntity.Kind.FULL) {
      // For now, always run full pipeline for FULL kind
      // This could be enhanced later with proper message context checking
      return ExecutionType.FULL_PIPELINE
    }

    // For new conversations, start with plan+maybe clarify
    return ExecutionType.PLAN_CLARIFY
  }

  private enum class ExecutionType {
    PLAN_CLARIFY,
    FULL_PIPELINE,
    PARTIAL_REVISION
  }
}